src/Prefs.h

```diff
@@ -55,6 +57,168 @@ extern int gMenusDirty;
+      : CachingSettingBase< T >{ path }
+      , mDefaultValue{ defaultValue }
+   {}
+
+   //! This overload causes recomputation of the default each time it is needed
+   Setting( const SettingBase &path, DefaultValueFunction function )
+      : CachingSettingBase< T >{ path }
+      , mFunction{ function }
+   {}
+   
+
+   const T& GetDefault() const
+   {
+      if ( mFunction )
+         mDefaultValue = mFunction();
+      return mDefaultValue;
+   }
+
+   //! overload of Read returning a success flag
+   /*! The cache is unchanged if that is false */
+   bool Read( T *pVar ) const
+   {
+      if ( pVar && this->mValid ) {
+         *pVar = this->mCurrentValue;
+         return true;
+      }
+      const auto config = this->GetConfig();
+      if ( pVar && config ) {
+         if ((this->mValid = config->Read( this->mPath, &this->mCurrentValue )))
+            *pVar = this->mCurrentValue;
+         return this->mValid;
+      }
+      return (this->mValid = false);
+   }
+
+   //! overload of Read, always returning a value
+   /*! The value is the default stored in this in case the key is known to be absent from the config;
+    but it returns type T's default value if there was failure to read the config */
+   T Read() const
+   {
+      return ReadWithDefault( GetDefault() );
+   }
+
+   //! new direct use is discouraged but it may be needed in legacy code
+   /*! Use the given default in case the preference is not defined, which may not be the
+    default-default stored in this object. */
+   T ReadWithDefault( const T &defaultValue ) const
+   {
+      const auto config = this->GetConfig();
+      return config
+         ? ( this->mValid = true, this->mCurrentValue =
+               config->ReadObject( this->mPath, defaultValue ) )
+         : T{};
+   }
+
+   //! Write value to config and return true if successful
+   bool Write( const T &value )
+   {
+      const auto config = this->GetConfig();
+      if ( config ) {
+         this->mCurrentValue = value;
+         return DoWrite();
+      }
+      return false;
+   }
+
+   //! Reset to the default value
+   bool Reset()
+   {
+      return Write( GetDefault() );
+   }
+
+protected:
+   //! Write cached value to config and return true if successful
+   /*! (But the config object is not flushed) */
+   bool DoWrite( )
+   {
+      const auto config = this->GetConfig();
+      return this->mValid =
+         config ? config->Write( this->mPath, this->mCurrentValue ) : false;
+   }
+
+   mutable T mDefaultValue{};
+   const DefaultValueFunction mFunction;
+};
+
+//! This specialization of Setting for bool adds a Toggle method to negate the saved value
+class BoolSetting final : public Setting< bool >
+{
+public:
+   using Setting::Setting;
+
+   //! Write the negation of the previous value, and then return the current value.
+   bool Toggle();
+};
+
+//! Specialization of Setting for int
+class IntSetting final : public Setting< int >
```

Comment:
Not that it really matters, but feels like `using IntSetting = Setting<int>;` looks cleaner in this context

Reply:
I needed to derive BoolSetting to add a method.  Doing this for the other specializations just looked more consistent.  Not very important.

Reply:
Another advantage of doing it as I did, is that it's easier to forward-declare IntSetting etc. as incomplete types.